runtime/common/src/helixbridge/mod.rs

```diff
@@ -0,0 +1,446 @@
+	ensure,
+	pallet_prelude::*,
+	traits::{Currency, ExistenceRequirement, Get, WithdrawReasons},
+	transactional,
+	weights::PostDispatchInfo,
+	PalletId,
+};
+use frame_system::{ensure_signed, RawOrigin};
+use sp_core::H256;
+use sp_runtime::{
+	traits::{AccountIdConversion, BadOrigin, CheckedDiv, CheckedMul, Convert, Saturating, Zero},
+	DispatchErrorWithPostInfo, MultiSignature, MultiSigner,
+};
+use sp_std::{str, vec, vec::Vec};
+
+pub use pallet::*;
+pub type ChainName = Vec<u8>;
+pub type AccountId<T> = <T as frame_system::Config>::AccountId;
+pub type RingBalance<T> = <<T as Config>::RingCurrency as Currency<AccountId<T>>>::Balance;
+
+/// The parameters box for the pallet runtime call.
+#[derive(Encode, Decode, Debug, PartialEq, Eq, Clone)]
+pub enum CallParams<T: Config> {
+	#[codec(index = 1)]
+	S2sBackingPalletUnlockFromRemote(RingBalance<T>, AccountId<T>),
+}
+/// Creating a concrete message payload which would be relay to target chain.
+pub trait CreatePayload<
+	SourceChainAccountId,
+	TargetChainAccountPublic,
+	TargetChainSignature,
+	T: Config,
+>
+{
+	type Payload: Encode;
+
+	fn encode_call(pallet_index: u8, call_params: CallParams<T>) -> Result<Vec<u8>, &'static str> {
+		let mut encoded = vec![pallet_index];
+		encoded.append(&mut call_params.encode());
+		Ok(encoded)
+	}
+
+	fn create(
+		origin: CallOrigin<SourceChainAccountId, TargetChainAccountPublic, TargetChainSignature>,
+		spec_version: u32,
+		weight: u64,
+		call_params: CallParams<T>,
+		dispatch_fee_payment: DispatchFeePayment,
+	) -> Result<Self::Payload, &'static str>;
+}
+
+pub trait LatestMessageNoncer {
+	fn outbound_latest_generated_nonce(lane_id: LaneId) -> u64;
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::config]
+	#[pallet::disable_frame_system_supertrait_check]
+	pub trait Config: frame_system::Config {
+		/// The pallet id of this pallet
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+
+		/// The *RING* currency.
+		type RingCurrency: Currency<AccountId<Self>>;
+
+		/// The bridge account id converter.
+		/// `remote account` + `remote chain id` derive the new account
+		type BridgedAccountIdConverter: Convert<H256, Self::AccountId>;
+
+		/// The bridged chain id
+		type BridgedChainId: Get<ChainId>;
+
+		/// Outbound payload creator used for s2s message
+		type OutboundPayloadCreator: Parameter
+			+ CreatePayload<Self::AccountId, MultiSigner, MultiSignature, Self>;
+
+		/// The remote chain name where the backing module in
+		type BackingChainName: Get<ChainName>;
+
+		/// The lane id of the s2s bridge
+		type MessageLaneId: Get<LaneId>;
+
+		/// The local token decimals.
+		#[pallet::constant]
+		type DecimalsDifference: Get<RingBalance<Self>>;
```

Comment:
```suggestion
		type DecimalMultiplier: Get<u128>;
```
I'm not sure this multiplier's type should be `RingBalance ` or not.

I think it should be u128

Reply:
Defining decimals using u128 is more accurate, but the operations between Balance and u128 still requires a type conversion. So I'm not sure which type is better.